### PR TITLE
Add swift:4.1 with Codable to default set of runtimes

### LINKED ISSUE #3414: Add swift 4.1 runtime/kind to default set of runtimes.
- [ ] Add runtime image as part of gradle build
- [ ] Update runtime manifest in ansible (default swift 4.1)
- [ ] Add tests cases for swfit 4.1 
- [ ] Update docs 
  - [ ] actions.md for the new kind
  - [ ] reference.md add docs on how to migrate from swift 3 to 4

### PR DESCRIPTION
**Pull request by csantanapr**
Closes #3414 

## Description
Add swift:4.1 kind for default set of runtimes.
Swift 4.x adds new signature using async callable function and JSON sed using Codable.
The runtime doesn't provide any contraints on packages to include, it doesn't come with any external dependencies like SwiftlyJSON.

## Related issue and scope
- [x] Issue #3414 

## My changes affect the following components
- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [x] Tests
- [x] Deployment
- [ ] CLI
- [ ] General tooling
- [x] Documentation
- [x] Runtimes

## Types of changes
- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:
- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [x] I added tests to cover my changes.
- [x] My changes require further changes to the documentation.
- [x] I updated the documentation where necessary.

### CODE REVIEW

**Comment by csantanapr** [comment]
TODOs
- [ ] Update Docs

**Comment by csantanapr** [comment]
## Swift 4.x support

Some examples of using Codable In and Out
### Codable style function signature
Create file `helloCodableAsync.swift`
```swift
// Domain model/entity
struct Employee: Codable {
  let id: Int?
  let name: String?
}
// codable main function
func main(input: Employee, respondWith: (Employee?, Error?) -> Void) -> Void {
    // For simplicity, just passing same Employee instance forward
    respondWith(input, nil)
}
```
```
wsk action update helloCodableAsync helloCodableAsync.swift swift:4.1
```
ok: updated action helloCodableAsync
```
wsk action invoke helloCodableAsync -r -p id 42 -p name Carlos
```
```json
{
    "id": 42,
    "name": "Carlos"
}
```

### Codable Error Handling
Create file `helloCodableAsync.swift`
```swift
struct Employee: Codable {
    let id: Int?
    let name: String?
}
enum VendingMachineError: Error {
    case invalidSelection
    case insufficientFunds(coinsNeeded: Int)
    case outOfStock
}
func main(input: Employee, respondWith: (Employee?, Error?) -> Void) -> Void {
    // Return real error
    do{
        throw VendingMachineError.insufficientFunds(coinsNeeded: 5)
    } catch {
        respondWith(nil, error)
    } 
}
```
```
wsk action update helloCodableError helloCodableError.swift swift:4.1
```
ok: updated action helloCodableError
```
wsk action invoke helloCodableError -b -p id 42 -p name Carlos
```
```json
{
"name": "helloCodableError",
"response": {
  "result": {
    "error": "insufficientFunds(5)"
  },
"status": "application error",
"success": false
}
```

### Packaging an action as a Swift executable using Swift 4

When you create an OpenWhisk Swift action with a Swift source file, it has to be compiled into a binary before the action is run. Once done, subsequent calls to the action are much faster until the container holding your action is purged. This delay is known as the cold-start delay.

To avoid the cold-start delay, you can compile your Swift file into a binary and then upload to OpenWhisk in a zip file. As you need the OpenWhisk scaffolding, the easiest way to create the binary is to build it within the same environment as it will be run in. These are the steps:

- Run an interactive Swift action container.
  ```
  docker run --rm -it -v "$(pwd):/owexec" openwhisk/action-swift-v4.0 bash
  ```
  This puts you in a bash shell within the Docker container.

- Copy the source code and prepare to build it.
  ```
  cp /owexec/hello.swift /swift4Action/spm-build/Sources/Action/main.swift
  ```
  ```
  cat /swift4Action/epilogue.swift >> /swift4Action/spm-build/Sources/Action/main.swift
  ```
  ```
  echo '_run_main(mainFunction:main)' >> /swift4Action/spm-build/Sources/Action/main.swift
  ```
  Copy any additional source files to `/swift4Action/spm-build/Sources/Action/`


- (Optional) Create the `Package.swift` file to add dependencies.
```swift
// swift-tools-version:4.0
// The swift-tools-version declares the minimum version of Swift required to build this package.

import PackageDescription

let package = Package(
    name: "Action",
    products: [
      .executable(
        name: "Action",
        targets:  ["Action"]
      )
    ],
    dependencies: [
      .package(url: "https://github.com/IBM-Swift/SwiftyRequest.git", .upToNextMajor(from: "1.0.0"))
    ],
    targets: [
      .target(
        name: "Action",
        dependencies: ["SwiftyRequest"],
        path: "."
      )
 
```
  As you can see this example adds `SwiftyRequest` dependencies.
  
  Notice that now with swift:4.0 is no longer required to include `CCurl`, `Kitura-net` and `SwiftyJSON` in your own `Package.swift`.
  You are free now to use no dependencies, or add the combination that you want with the versions you want.

- Copy Package.swift to spm-build directory
  ```
  cp /owexec/Package.swift /swift4Action/spm-build/Package.swift
  ```

- Change to the spm-build directory.
  ```
  cd /swift4Action/spm-build
  ```

- Compile your Swift Action.
  ```
  swift build -c release
  ```

- Create the zip archive.
  ```
  zip /owexec/hello.zip .build/release/Action
  ```

- Exit the Docker container.
  ```
  exit
  ```

  This has created hello.zip in the same directory as hello.swift.

- Upload it to OpenWhisk with the action name helloSwifty:
  ```
  wsk action update helloSwiftly hello.zip openwhisk/action-swift-v4.0
  ```

- To check how much faster it is, run
  ```
  wsk action invoke helloSwiftly --blocking
  ```

### Migrating from Swift 3 to Swift 4

### Helper compile.sh helper script
When compiling and packaging your swift 4 action, there are a couple of differences.
All your source code needs to be copied to `/swift4Action/spm-build/Sources/Action/` instead of `/swift3Action/spm-build/`
You Package.swift needs to have the first line with a comment indicating swift4 tooling and format
```
// swift-tools-version:4.0
```
For swift 4 you need specify additional information in Package.swift such as `products` with executable name `Action` and `targets`

You can take a look at the helper script [tools/build/compile.sh](tools/build/compile.sh) to compile and zip your Actions.
Having a project directory `Hello` under a directory `actions` like the following:
```
actions/Hello/Package.swift
actions/Hello/Sources/main.swift
```
Change to the parent directory then run the compile script specify the project directory, the kind `swift:3.1.1` or `swift:4.0` and any swiftc build flags like the following:
```
cd actions/
incubator-runtime-swift/tools/build/compile.sh Hello swift:4.0 -v
```
This will produce a zip `build/swift4/Hello.zip`

### SwiftyJSON using single source action file
If you have a swift:3.1.1 action not compile, just as source using the `SwiftyJSON` package, you need to precompile your action and specify the version of SwiftyJSON you wan to use for swift:4.0 kind action.
Take into account that starting with Swift 4 there is better support to manage JSON data natively.

Note: This is only applicable to the base image provided for the Swift 4 runtime, other downstream such as IBM Cloud Functions extending this image might provide additional SDK and packages including `SwiftyJSON` and IBM Watson SDK, check the vendor documentation for more specific information about packages and versions.

### Building the Swift4 Image
```
./gradlew core:swift40Action:distDocker
```
This will produce the image `whisk/action-swift-v4.0`

Build and Push image
```
docker login
./gradlew core:swift40Action:distDocker -PdockerImagePrefix=$prefix-user -PdockerRegistry=docker.io 
```

### Swift 4.1 Experimental
We have a runtime for swift 4.1, is experimental as we are trying beta builds released by Swift org.
Follow same insructions for Swift 4.0 above and replace the kind wih `swift:4.1` and image with `openwhisk/action-swift-v4.0`

**Comment by csantanapr** [comment]
PG4/1469 🔵

**Comment by csantanapr** [comment]
added initial stab at docs

File: docs/actions.md [hunk 1]
```diff
@@ -685,17 +686,45 @@ func main(args: [String:Any]) -> [String:Any] {
     }
 }
 ```
+In this example the Swift action consume a dictionary and produce a dictionary.
```
Comment (rabbah) [inline docs/actions.md]: s/consume/consumes/
s/produce/produces/

File: docs/actions.md [hunk 2]
```diff
@@ -685,17 +686,45 @@ func main(args: [String:Any]) -> [String:Any] {
...
+
+You can create a OpenWhisk action called `helloSwift` from this function as
```
Comment (rabbah) [inline docs/actions.md]: s/a/an/

File: docs/reference.md [hunk 3]
```diff
@@ -378,13 +378,19 @@ Python 2 actions are executed using Python 2.7.12. This is the default runtime f
...
 Swift 3 actions are executed using Swift 3.1.1  `--kind swift:3.1.1`.  
-The default `--kind swift:default` is Swift 3.1.1.
 
 Swift 3.1.1 actions can use the following packages:
 - KituraNet version 1.7.6, https://github.com/IBM-Swift/Kitura-net
 - SwiftyJSON version 15.0.1, https://github.com/IBM-Swift/SwiftyJSON
 - Watson Developer Cloud SDK version 0.16.0, https://github.com/watson-developer-cloud/swift-sdk
 
+### Swift 4
+Swift 3 actions are executed using Swift 4.1  `--kind swift:4.1`.
```
Comment (rabbah) [inline docs/reference.md]: swift 4 actions?

**Comment by dubeejw** [comment]
PG1 2713 ⏳

**Comment by csantanapr** [comment]
Playground1/2713 🔵

**Pull request merged**